src/tri/_tri.h

```diff
@@ -161,12 +163,12 @@ void write_contour(const Contour& contour);
 class Triangulation
 {
 public:
-    typedef numpy::array_view<const double, 1> CoordinateArray;
-    typedef numpy::array_view<double, 2> TwoCoordinateArray;
-    typedef numpy::array_view<int, 2> TriangleArray;
-    typedef numpy::array_view<const bool, 1> MaskArray;
-    typedef numpy::array_view<int, 2> EdgeArray;
-    typedef numpy::array_view<int, 2> NeighborArray;
+    typedef py::array_t<double, py::array::c_style | py::array::forcecast> CoordinateArray;
+    typedef py::array_t<double, py::array::c_style | py::array::forcecast> TwoCoordinateArray;
+    typedef py::array_t<int,    py::array::c_style | py::array::forcecast> TriangleArray;
+    typedef py::array_t<bool,   py::array::c_style | py::array::forcecast> MaskArray;
+    typedef py::array_t<int,    py::array::c_style | py::array::forcecast> EdgeArray;
+    typedef py::array_t<int,    py::array::c_style | py::array::forcecast> NeighborArray;
```

Comment:
Trying to learn a bit... 1 is c_style, but 2 is f_style as I understand it. I assume that it works either way, but is there a reason that 1 or 2 was originally selected? (I do not get the difference to start with...)

Reply:
The previous `array_view` class was templated on both `dtype` and `ndim` whereas the `pybind11::array_t` class is only templated on `dtype`.

The `pybind11` flags are required for passing arrays from Python to C++ to ensure that we receive numpy arrays of the correct `dtype` and order in memory (https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#arrays).

`py::array_t::c_style` is C-ordered (row major) rather than fortran-ordered (column major). `py::array::forcecast` forces a cast to the required `dtype` if it is not already so. The combination is the equivalent of the Python NumPy function `np.ascontiguousarray(input_array, dtype=whatever)`. Combining the two means that on the C++ side an array is a contiguous buffer in memory which we can walk using pointer arithmetic in the conventional manner.

The C-style flag isn't really needed for 1D arrays. The C++ implementation could have been written for fortran-ordered arrays :grimacing:.

Reply:
Ahh, thanks for clarifying! I guessed f was for Fortran, but couldn't really understand how they could differ as I didn't think beyond 1D...